src/ynEIGEN/TokenStakingNode.sol

```diff
@@ -135,105 +142,152 @@ contract TokenStakingNode is
+                revert WithdrawalAmountMismatch();
+            }
+
+            queuedShares[_strategy] -= withdrawnShares;
         }
 
-        uint256 _actualAmountOut = _token.balanceOf(address(this)) - _balanceBefore;
-        IWrapper _wrapper = IYieldNestStrategyManager(tokenStakingNodesManager.yieldNestStrategyManager()).wrapper();
-        IERC20(_token).forceApprove(address(_wrapper), _actualAmountOut); // NOTE: approving also token that will not be transferred
-        (_actualAmountOut, _token) = _wrapper.wrap(_actualAmountOut, _token);
+        uint256 withdrawnBalance = token.balanceOf(address(this)) - oldBalance;
+        
+        IWrapper wrapper = IYieldNestStrategyManager(tokenStakingNodesManager.yieldNestStrategyManager()).wrapper();
+        token.forceApprove(address(wrapper), withdrawnBalance); // NOTE: approving also token that will not be transferred
+        (withdrawnBalance, token) = wrapper.wrap(withdrawnBalance, token);
 
-        queuedShares[_strategy] -= _shares;
-        withdrawn[_token] += _actualAmountOut;
+        withdrawn[token] += withdrawnBalance;
 
-        if (updateTokenStakingNodesBalances) {
+        if (_updateTokenStakingNodesBalances) {
             // Actual balance changes only if slashing occured. choose to update here
             // only if the off-chain considers it necessary to save gas
-            IYieldNestStrategyManager(tokenStakingNodesManager.yieldNestStrategyManager()).updateTokenStakingNodesBalances(
-                _token
-            );
+            IYieldNestStrategyManager(tokenStakingNodesManager.yieldNestStrategyManager()).updateTokenStakingNodesBalances(token);
         }
 
-        emit CompletedQueuedWithdrawals(_shares, _actualAmountOut, address(_strategy));
+        emit CompletedQueuedWithdrawals(_scaledShares, withdrawnBalance, address(_strategy));
+    }
+
+    /**
+     * @notice Syncs queued shares in case of slashing to prevent accounting discrepancies between withdrawal queue and completion.
+     * @dev This function is to be called as soon as slashing is detected.
+     */
+    function syncQueuedShares() external onlyDelegator {
+        IDelegationManager delegationManager = tokenStakingNodesManager.delegationManager();
+
+        (IDelegationManagerTypes.Withdrawal[] memory withdrawals, uint256[][] memory withdrawableShares) = delegationManager.getQueuedWithdrawals(address(this));
+
+        for (uint256 i = 0; i < withdrawals.length; i++) {
+            IDelegationManagerTypes.Withdrawal memory withdrawal = withdrawals[i];
+
+            for (uint256 j = 0; j < withdrawal.strategies.length; j++) {
+                IStrategy strategy = withdrawal.strategies[j];
+                uint256 strategyWithdrawableShares = withdrawableShares[i][j];
+                bytes32 withdrawalRoot = delegationManager.calculateWithdrawalRoot(withdrawal);
+
+                queuedShares[strategy] = 0;
```

Comment:
this doesn't look right

assuming there are multiple withdrawals for the same strategy what was accumulated in the loop before that gets reset here

Reply:
You are right. Fixed it at https://github.com/yieldnest/yieldnest-protocol/pull/214/commits/ff24afc9cdcb195f6b19a4f9a88db29a85b655a8.

When writing the tests for this, I'll look for alternative implementations that might be cheaper and see if I can reduce the number of loops.